File: app/src/main/java/com/bookyrself/bookyrself/data/Profile/ProfileRepo.java
package com.bookyrself.bookyrself.data.Profile;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.bookyrself.bookyrself.data.ResponseModels.User.User;
import com.bookyrself.bookyrself.services.FirebaseService;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.ChildEventListener;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;

import io.reactivex.Flowable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

public class ProfileRepo implements ProfileDataSource {

    private DatabaseReference db;


    public ProfileRepo() {
        if (FirebaseAuth.getInstance().getUid() != null) {

            String userId = FirebaseAuth.getInstance().getUid();

            this.db = FirebaseDatabase.getInstance().getReference()
                    .child("users")
                    .child(userId);

            this.db.addChildEventListener(new ChildEventListener() {
                @Override
                public void onChildAdded(@NonNull DataSnapshot dataSnapshot, @Nullable String s) {
                    // Update repos observable to fetch new data
                    getProfileInfo(userId);
                }

                @Override
                public void onChildChanged(@NonNull DataSnapshot dataSnapshot, @Nullable String s) {
                    getProfileInfo(userId);
                }

                @Override
                public void onChildRemoved(@NonNull DataSnapshot dataSnapshot) {
                    getProfileInfo(userId);
                }

                @Override
                public void onChildMoved(@NonNull DataSnapshot dataSnapshot, @Nullable String s) {
                    getProfileInfo(userId);
                }

                @Override
                public void onCancelled(@NonNull DatabaseError databaseError) {
                    getProfileInfo(userId);
                }
            });
        }
     }


    @Override
    public Flowable<User> getProfileInfo(String userId) {

        return FirebaseService.getAPI()
                .getUserDetails(userId)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }

    @Override
    public Flowable<User> updateProfileInfo(String userId, User user) {
        return FirebaseService.getAPI()
                .patchUser(user,userId)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());

    }
}

File: app/src/main/java/com/bookyrself/bookyrself/data/profile/ProfileRepo.kt
package com.bookyrself.bookyrself.data.profile

import android.content.Context
import com.bookyrself.bookyrself.data.SingletonHolder
import com.bookyrself.bookyrself.data.profile.ProfileRepoResponse.Success
import com.bookyrself.bookyrself.data.profile.ProfileRepoResponse.Failure
import com.bookyrself.bookyrself.data.serverModels.User.User
import com.bookyrself.bookyrself.services.FirebaseServiceCoroutines
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.*

class ProfileRepo private constructor(context: Context) {

    private var db: DatabaseReference? = null
    private var cacheIsDirty: Boolean = true
    private var user: User? = null

    companion object : SingletonHolder<ProfileRepo, Context>(::ProfileRepo)

    init {
        FirebaseAuth.getInstance().uid?.let { userId ->
            this.db = FirebaseDatabase.getInstance().reference
                .child("users")
                .child(userId)

            this.db?.addChildEventListener(object : ChildEventListener {
                override fun onChildAdded(dataSnapshot: DataSnapshot, s: String?) {
                    // Update repos observable to fetch new data
                    cacheIsDirty = true
                }

                override fun onChildChanged(dataSnapshot: DataSnapshot, s: String?) {
                    cacheIsDirty = true
                }

                override fun onChildRemoved(dataSnapshot: DataSnapshot) {
                    cacheIsDirty = true
                }

                override fun onChildMoved(dataSnapshot: DataSnapshot, s: String?) {
                    cacheIsDirty = true
                }

                override fun onCancelled(databaseError: DatabaseError) {
                    cacheIsDirty = true
                }
            })
        }
    }


    suspend fun getProfileInfo(userId: String): ProfileRepoResponse {
        return if (cacheIsDirty) {
            val profileResponse = FirebaseServiceCoroutines.instance.getUserDetails(userId)
            if (profileResponse.isSuccessful) {
                user = profileResponse.body()
                Success(user = profileResponse.body())
            } else {
                Failure(errorMessage = profileResponse.message())
            }
        } else {
            Success(user)
        }
    }

    suspend fun updateProfileInfo(userId: String, user: User): ProfileRepoResponse {
        val response = FirebaseServiceCoroutines.instance.patchUser(user, userId)
        return if (response.isSuccessful) {
            ProfileRepoResponse.Success(response.body())
        } else {
            ProfileRepoResponse.Failure(response.message())
        }
    }
}


sealed class ProfileRepoResponse {
    class Success(val user: User?) : ProfileRepoResponse()
    class Failure(val errorMessage: String) : ProfileRepoResponse()
}
